bridge-fungible-token/src/utils.sw

```diff
@@ -1,10 +1,161 @@
 library utils;
 
-use std::mem::read;
+dep errors;
+dep events;
+dep data;
+
+use std::{
+    constants::ZERO_B256,
+    flags::{
+        disable_panic_on_overflow,
+        enable_panic_on_overflow
+    },
+    math::*,
+    vec::Vec
+};
+
+use errors::BridgeFungibleTokenError;
+use data::MessageData;
+
+// the function selector for finalizeWithdrawal on the L1ERC20Gateway contract
+const FINALIZE_WITHDRAWAL_SELECTOR: u64 = 0x53ef1461;
 
 // TODO: [std-lib] remove once standard library functions have been added
 const GTF_INPUT_MESSAGE_DATA_LENGTH = 0x11B;
 const GTF_INPUT_MESSAGE_DATA = 0x11E;
+const GTF_INPUT_MESSAGE_SENDER = 0x115;
+const GTF_INPUT_MESSAGE_RECIPIENT = 0x116;
+
+fn decimal_adjustment_factor() -> u64 {
+    if LAYER_1_DECIMALS > DECIMALS {
+        10.pow(LAYER_1_DECIMALS - DECIMALS)
+    } else if LAYER_1_DECIMALS < DECIMALS {
+        // TODO: Discuss how we want to handle this case
+        1
+    } else {
+        1
+    }
+}
+
+/// used to increase the amount of "decimal" places by appending the
+/// appropriate amount of 0s to the u64 via multiplication by the appropriate
+/// adjustment_factor.
+/// Potential overflow is accounted for & the result is returned as a b256
+pub fn safe_u64_to_b256(val: u64) -> b256 {
+    let adjustment_factor = decimal_adjustment_factor();
+    let mut result: b256 = ZERO_B256;
+    disable_panic_on_overflow();
+    asm(product, overflow, value: val, factor: adjustment_factor, ptr: __addr_of(result)) {
+        mul product value factor;
+        move overflow of;
+        sw ptr product i3;
+        sw ptr overflow i2;
+    }
+    enable_panic_on_overflow();
+    result
+}
+
+pub fn safe_b256_to_u64(val: b256) -> Result<u64, BridgeFungibleTokenError> {
+    // first, decompose into u64 values
+    let u64s = decompose(val);
+    let adjustment_factor = decimal_adjustment_factor();
+
+    // verify amount will require no partial refund of dust by ensuring that
+    // the first n decimal places in the passed-in value are empty,
+    // then verify amount is not too small or too large
+    if (u64s.3 / adjustment_factor) * adjustment_factor == u64s.3
+        && u64s.3 >= adjustment_factor
+        && u64s.0 == 0
+        && u64s.1 == 0
+        && u64s.2 == 0
+    {
+        // reduce decimals
+        Result::Ok(u64s.3 / adjustment_factor)
```

Comment:
I forgot about `U256`. We could use that to do the appropriate math. Basically if a modulo operation with the "adjustment_factor" returns anything other than zero, then the function should fail. Also if after dividing by the "adjustment_factor", the value is larger than what u64 can hold, then the function should fail.

It might be worth it to make everywhere that a `b256` is currently used to represent amounts, we switch it to `U256`, rather than just trying to cast to `U256` in this function alone.

Reply:
My only concern with the U256 approach is that there was a lengthy discussion about it in the language sync on Monday. Some users are finding that the implementation is not optimized enough and this is causing very significant bytecode bloat and out-of-gas errors. This is at least partly due to the fact that the FuelVM uses 64 bit registers. One recent example on slack describes an amm implemented in Sway which clones a Solidity project.
The Solidity is ~25KiB, and the Sway is 5MB. Compiler optimization passes haven't been implemented yet, so it's expected that this will get much better.

Reply:
Another possibility to to extend our existing `b256` lib.
I've already implemented bitwise ops for `b256` in `core::ops`.
I could look add adding the remaining needed operations by leveraging the existing suite of bitwise operators / bitshifting as per this resource: https://bisqwit.iki.fi/story/howto/bitmath/

Reply:
I was looking into manually implementing division last night before I realized U256 already had the common "subtract and shift" algorithm implemented which can get us through a first draft. There is a potentially faster algorithm for our use case though since our divisor is not also a large u256 number, but I didn't fully get it figured out. We can make an issue and explore it further, but for now I think it would be best to utilize the U256 math (although maybe we don't change `b256` to `U256` everywhere in the code just yet)

Reply:
Also, I'm not sure it's appropriate to add math operation (beyond just bitwise) to `b256` since that type is just representative of bits, not a number. The bits could be a signed int, unsigned int, some kind of floating point number. We don't know.

Reply:
Fair points. I will switch to using the `U256` type where we really need it, and go over the contract once more for places like this one where assumptions are made about the relative size of the L1 & L2 decimal values.

Reply:
Changes made.